OpenAPI schemas

traines-source opened the pull request:
* fixes pretty param to be a query param
* adds missing from/to params for /journeys
* adds some response schemas

disclaimer:
* even fields marked as required or not nullable might be missing or null under unknown circumstances
* some fields not marked as required might always be present
* some fields present in some responses are missing from the schema
--> don't rely on it, just use it as a quick way to generate a client

derhuerst commented on lib/json-pretty-printing.js:
```diff
@@ -10,7 +10,7 @@ const configureJSONPrettyPrinting = (req, res) => {
 const jsonPrettyPrintingOpenapiParam = {
 	name: 'pretty',
-	in: 'path',
+	in: 'query',
```
Good catch, thanks!

derhuerst commented:
When we chatted before, I forgot to tell you that there are [community-maintained TypeScript typings](https://www.npmjs.com/package/@types/hafas-client), sorry. These may (have been) used to generate the JSON Schema definitions.

derhuerst commented on routes/arrivals.js:
```diff
@@ -147,8 +147,284 @@ Works like \`/stops/{id}/departures\`, except that it uses [\`hafasClient.arriva
 								schema: {
-									type: 'array',
-									items: {type: 'object'}, // todo
+									'type': 'array',
```
What do you think about moving the each "root" schema into a variable, to reduce the nesting level?

derhuerst commented on routes/arrivals.js:
```diff
@@ -147,8 +147,284 @@ Works like \`/stops/{id}/departures\`, except that it uses [\`hafasClient.arriva
+											'tripId': {
+												'type': 'string'
+											},
+											'stop': {
```
Let's move re-used sub schemas (e.g. a stop) into a variable each, and then use them here.

traines-source commented:
Auto-generating the schemas from the TS typings instead of from some JSON instances indeed seems better. There is only one drawback I have noticed so far: It doesn't distinguish between null and not present. But most libraries/programming languages don't do that anyways, so it shouldn't really be a problem.

Thinking about it again, we should have one common schema file for all endpoints. That way, it's easier to replace it with a new auto-generated version. And the reused sub-schemas (using "$ref") actually get reused and turn into the same class in the generated code in the client. I will have to take a more thorough look at this.

derhuerst commented:
> Thinking about it again, we should have one common schema file for all endpoints. That way, it's easier to replace it with a new auto-generated version. And the reused sub-schemas (using "$ref") actually get reused and turn into the same class in the generated code in the client. I will have to take a more thorough look at this.

We can host the schemas in the `hafas-rest-api#4` branch, and then re-use them in `bvg-rest#6`, `db-rest#6`, `hvv-rest#6` & `vbb-rest#6`. What do you think?

traines-source commented:
In case that didn't become clear, I was talking about the endpoints like journeys, departures, arrivals, etc., which in the first iteration have completely separate schemas. In a generated client, that would lead to separate classes of the same structure (e.g. for a stop).
I had hoped that the schemas propagate from hafas-rest-api to db-rest etc., but I haven't tested it.

derhuerst commented:
> In case that didn't become clear, I was talking about the endpoints like journeys, departures, arrivals, etc., which in the first iteration have completely separate schemas.

Yes, we can refactor common parts into common (sub-)schemas later.

> I had hoped that the schemas propagate from hafas-rest-api to db-rest etc., but I haven't tested it.

They do, but some `*-rest` projects have a slightly customised response, given that they use a [`hafas-client` profile](https://github.com/public-transport/hafas-client/tree/5.25.0/p) that modifies the data parsed from HAFAS.

traines-source commented:
I have changed the approach now.
- use one central schema to reuse ($ref) definitions between endpoints, so that only one class per type is generated in clients
- generate schema file from @types/hafas-client using script (using JSONSchema lib, some alterations for compatibility with OpenAPI necessary)
- no fields marked as required

derhuerst commented:
I finally found time and energy to look at this from a broader perspective. Thank you for waiting.

Also, I'd like to cc @bergmannjg, who has contributed the `hafas-client` typings, and generally uses and knows about typings a lot more in this context. Do you have an opinion on how we should derive the OpenAPI/JSON Schema definitions from a `hafas-client`?

Some problems that are not yet solved by this PR:

## profiles

`hafas-client` provides a common API to query *several* HAFAS endpoints; It also provides endpoint-specific customisations called "profiles", each with
- a custom list of products
- minor (but relevant) tweaks to `hafas-client`'s interface.

For example, the `hafas-client/p/db` profile adds an `opt.firstClass: boolean` flag to `journeys()` (input), and a `loadFactor` field to `Arrival` & `Departure` (output).

As the enum of supported products is quite relevant to people using `hafas-rest-api`-based APIs, and as profiles might actually alter `hafas-client`'s interface in a backwards-incompatible way, it is **important to generate *custom* OpenAPI/JSON Schema definitions for each `hafas-rest-api`-consuming project**.

@bergmannjg As a first step towards this, do you think it would be possible to let `@types/hafas-client/p/*/index.d.ts` define and export a modified/extended TypeScript schema definition that not only describes what the respective profile looks like, but also what the respective `createClient(profile)` will look like? For example, `@types/hafas-client/p/db/index.d.ts` would export a `HafasClientWithDbProfile` interface *with* `radar()` and  *without* `lines()`. I don't know the TypeScript tooling well enough to answer if this is feasible.

As the next step, we could modify `scripts/generate-schema.js` from this PR to

1. take either the path to a TypeScript schema definition (preferred) or the name of a `hafas-client` profile (less flexible) as an argument, and
2. generate the *profile-specific* OpenAPI definitions this way.

Dependents of `hafas-rest-api` (e.g. `db-rest`) would have to call `generate-schema.js` and and use the generated definitions when building the API server, just like how they do it with the Markdown API already ([1](https://github.com/derhuerst/db-rest/blob/5fc72f098e2a7b55f775ec9b1b87703f60de0d84/api-docs.js#L421-L425), [2](https://github.com/derhuerst/db-rest/blob/5fc72f098e2a7b55f775ec9b1b87703f60de0d84/api.js#L85-L87)).

## passing the generated defintions into `hafas-rest-api`

This follows from the requirement to generate `hafas-rest-api`-consumer-specific OpenAPI definitions (see above): The generated schema should be stored as a file in the consumer's directory, and passed into `hafas-rest-api`'s `createApi()`, e.g. as `openApiComponents`.

Then, all default routes (those implemented by `hafas-rest-api`), as well as those defined by the consumer, can access the HAFAS-endpoint-specific OpenAPI definitions.

I'm hestitant to pass a "full"/root OpenAPI spec into `createApi()` though, as

- it is not complete anyways, as the routes are missing, and
- it seems more cumbersome to customize than a plain `components` object.

What do you think?

## excluding `createClient.*`

Currently, the generated OpenAPI defintions contain (almost) every type/interface twice, as `Foo` and `createClient.Foo`. This is because the TypeScript typings (understandably) also describe `createClient()` & `createClient.Profile`, and apparently because `typescript-json-schema` generates re-used types/interfaces twice?

This is a minor issue, as it is only about noise/bloat being displayed in the OpenAPI GUIs and possibly generated in not-so-clever client generators, but nevertheless I think we should tackle it eventually™️.

related: https://github.com/YousefED/typescript-json-schema/issues/282

derhuerst commented:
Another unrelated note:
@traines-source Since you opened the PR, I have updated `master` to contain the `v6` implementation of the API(s), doing some backwards-incompatible changes and upgrading to `hafas-client@6`. You can either
- keep this PR against the `v5` code base and let me port it to `v6` (in this case, pick `5` as this PR's target branch), or
- adapt this PR to the `v6` changes.

The former option is probably less work for you, the latter less work for me. Up to you!

bergmannjg commented:
@derhuerst: a first note to your questions

> Do you have an opinion on how we should derive the OpenAPI/JSON Schema definitions from a hafas-client?

Please have look at [hafas-api-tsoa](https://github.com/bergmannjg/hafas-api-tsoa). It uses [tsoa](https://github.com/lukeautry/tsoa) as an alternative to *typescript-json-schema*. The api [spec](https://github.com/bergmannjg/hafas-api-tsoa/blob/master/src/swagger.json) is generated from the TS typings and from decorators in the [controller](https://github.com/bergmannjg/hafas-api-tsoa/blob/master/src/hafas/hafasController.ts) file.

>  It is important to generate custom OpenAPI/JSON Schema definitions for each hafas-rest-api consumer.

Usng tsoa the custom definitions may be generated from different controllers.

derhuerst commented:
I'm a bit hesitant to use tools like tsoa because they impose a very specific structure upon your API, which IMHO is neither more understandable (the controller pattern really doesn't fit to what `hafas-rest-api` does) nor more extendable.

derhuerst commented:
> @bergmannjg As a first step towards this, do you think it would be possible to let `@types/hafas-client/p/*/index.d.ts` define and export a modified/extended TypeScript schema definition that not only describes what the respective profile looks like, but also what the respective `createClient(profile)` will look like? For example, `@types/hafas-client/p/db/index.d.ts` would export a `HafasClientWithDbProfile` interface _with_ `radar()` and _without_ `lines()`. I don't know the TypeScript tooling well enough to answer if this is feasible.

Currently, `@types/hafas-client` describes an interface that is only given if I use `hafas-client/p/db` (not with e.g. `hafas-client/p/nvv`). Where shall I open an Issue so we can discuss this further? I would like to avoid (domain-specific) spam in the (very general-purpose) `DefinitelyTyped` repo.

bergmannjg commented:
This [branch](https://github.com/bergmannjg/hafas-client/tree/add-types-in-jsdoc) contains code to check and create the typing file, here is a short [description](https://github.com/bergmannjg/hafas-client/blob/add-types-in-jsdoc/typescript-readme.md).

>Currently, @types/hafas-client describes an interface that is only given if I use hafas-client/p/db (not with e.g. hafas-client/p/nvv)

Can you give an example, the types should work with all profiles.

traines-source commented:
Some thoughts:

* There was something going on why I needed both `createClient.Foo` and `Foo`, but I don't remember why, have to look at it again. Or maybe it was just that I was happy that it worked at all after having tried other JSON schema libs that didn't do what we need.
* I personally like the tsoa/in-code-annotation approach a lot, but I'm not sure it is easily applicable to the hafas-client codebase, that's probably what you meant @derhuerst. (But it saves you from the ugly string manipulation that I had to do to convert JSON schema to OpenAPI schema because I haven't found a lib that actually does that directly.)
* I feel that it is not absolutely necessary to take into account the different profiles (but I might not know them well enough). E.g. AFAIK the products are currently just represented as an arbitrary key/value object or string and loadFactor is just an optional field in the TS typings. For me, the OpenAPI spec is a "best-effort" thing that you might want to tweak to fit your needs anyways (anecdote: the lib that I'm using to create Golang types from the OpenAPI (2.0) spec generates nullable types for required fields and non-nullable types for non-required fields, with the justification that with required fields, it is important to know whether they were actively set, while for non-required fields the type default is fine, see [issue](https://github.com/go-swagger/go-swagger/issues/2056). For fields like `delay` where there is an important difference between `null` and `0`, I needed it the other way round and thus ironically I set it to required...)
* I already noticed some of the changes in v6 recently, should not be too much work I guess. But that is something that should be adjusted in the TS typings first.

derhuerst commented:
> * I personally like the tsoa/in-code-annotation approach a lot, but I'm not sure it is easily applicable to the hafas-client codebase, that's probably what you meant @derhuerst. (But it saves you from the ugly string manipulation that I had to do to convert JSON schema to OpenAPI schema because I haven't found a lib that actually does that directly.)

Yes, indeed it prevents the (somewhat hacky and potentially incomplete) manipulation of the generated JSON schema, at a high cost in other area: Structuring all of the code according to the controller pattern, which IMO does not fit well to `hafas-client`'s "RPC nature" as well as `hafas-rest-api`'s dynamic-ness.

> * I feel that it is not absolutely necessary to take into account the different profiles (but I might not know them well enough). […] For me, the OpenAPI spec is a "best-effort" thing that you might want to tweak to fit your needs anyways […].

Maybe I misunderstood you, but there are at least two reasons why I'm not a fan of this approach which both come down to long-term maintenance:

- With a) so many projects in general and b) so many stacked on top of `hafas-client`, I try to reduce both manual maintenance and human error caused by it to a minimum. This means that manually tweaking a schema (or double-checking a diff) after regenerating it is too error-prone.
	- (I'm aware that static typing can help with this *to a certain degree*, but so far, I don't see a way in which it doesn't get in the way of `hafas-client`'s (and all dependent libs') highly dynamic nature: Letting its profiles override arbitrary parts of the `createClient()` API. Maybe there is a way, see https://github.com/public-transport/hafas-client/issues/238 for more discussion.)
- People get in touch with my rather often with questions about `*.transport.rest`. While that is fine – In fact, I really appreciate that they don't just ignore the issue and/or give up, but help me see broken and unintuitive aspects! –, I'd like to keep the amount of questions caused by *unnecessarily unintuitive* behaviour of the APIs.

> * I already noticed some of the changes in v6 recently, should not be too much work I guess. But that is something that should be adjusted in the TS typings first.

Great! Let's first discuss the steps forward here though, so that you don't waste time working on things that we might revise anyways.

bergmannjg commented:
> it is important to generate custom OpenAPI/JSON Schema definitions for each hafas-rest-api consumer

It may be possible to generate profile specific typings from the general @types/hafas-client/index.d.ts file using
the corresponding _products.js_ and _index.js_ files. These typings can be the input of _generate-schema.js._

For example: change the type of mode
```
    mode: 'train' | 'bus' | 'watercraft' | 'taxi' | 'gondola' | 'aircraft' | 'car' | 'bicycle' | 'walking';
}
```

with

```
    mode: 'train' | 'bus' | 'taxi';
}
```
 for the nvv profile using the products array.

The flags of the profile (like _trip_ or _radar_) can be used to adapt hafas-client's interface methods.

Maybe there should be flags for db profile specific fields _firstclass_, _loadfactor_, etc. 

The conversions  can be implemented with a shell script or node program.

> excluding createClient.*

This should be solved with ES modules. The createClient namespace doesn't exist anymore.

traines-source commented:
> With a) so many projects in general and b) so many stacked on top of hafas-client, I try to reduce both manual maintenance and human error caused by it to a minimum. This means that manually tweaking a schema (or double-checking a diff) after regenerating it is too error-prone.

With manually tweaking the schema, I meant the end user who uses it to generate the classes/types in whatever programming language they are using. Talking about schema generation in hafas-rest-api, I think it's indeed a good idea to do it on the fly at runtime (the only drawback there that we have to add the TS typings as a runtime dependency). I can have a look at that when we have agreed on how to proceed and as soon as the TS typings have been updated to match v6.

Since there still doesn't seem to be anything that really differs structurally, I'm still not convinced we should adapt the schema to profiles. This can even have the drawback for end users that they can not use the same (generated) code base for different HAFAS APIs.

traines-source commented:
Fast forward two years :see_no_evil:, I've updated the OpenAPI schema for v6.
* adding (more) missing request params without which the OpenAPI playground is basically unusable (e.g. for journeys, radar)
* generating generic response schemas based on `@types/hafas-client` ~~(I couldn't really distill the best way forward for profile-specific schemas based on our previous discussion, although there now exists e.g. a `JourneysOptionsDbProfile` in the typings. I still think a generic schema is better than none :)~~ ~~Profile-agnostic proposition: For the profile-specific products filter, I added an `additionalProperties` object with boolean properties, which is by default flattened to normal query params according to the [OpenAPI 3.0 spec](https://swagger.io/docs/specification/v3_0/serialization/). One could add an even more generic object allowing any values in order to cover all profile-specific params like `loyaltyCard`, `firstClass` etc. in an easy way.~~

traines-source commented:
Fast forward only a few months,
* generating the schema on the fly when starting the API server
* auto-generating the profile-specific products filter on the fly. The db-specific `loyaltyCard`, `firstClass` and `age` actually already work (i.e. are exposed in OpenAPI) as they are configured in [db-rest](https://github.com/derhuerst/db-rest/blob/6b6c1d6bdb6b41ea547d1fc6a59ac2d3a3e107e0/api.js#L89C1-L107C2). The response schemas do stay profile-agnostic, i.e. with `additionalProperties` for products and actually missing tickets. But I think this is acceptable.

I've now committed the (slightly adapted) output openapi.json/yaml to [db-vendo-client](https://github.com/public-transport/db-vendo-client) as well.

derhuerst commented:
> Since there still doesn't seem to be anything that really differs structurally, I'm still not convinced we should adapt the schema to profiles.

I just went through all `hafas-client` profiles.

- Most leave `hafas-client`'s default interface as-is, they merely configure protocol details (e.g. authentication) and define the set of products.
- Some profiles extend the schema/interface.
	- Several profiles parse more location details (e.g. additional IDs of stops).
	- The `db` profile did most customisations, but it is defunct now.
- AFAIK there is no profile that removes fields from the schema/interface!

This is why I changed my mind: I don't think it's good to have profile-specific typings for `hafas-client`. Nevertheless, `hafas-rest-api` should provide an API to customize the OpenAPI schema, as outlined above by @traines-source. 👍

> This can even have the drawback for end users that they can not use the same (generated) code base for different HAFAS APIs.

I think someone who has a complex-enough use case to warrant doing this will be okay using >1 generated API clients.

derhuerst commented:
@bergmannjg FYI: Currently, in the `hafas-client` typings, `arrivals()` & `departures()` are specified as returning an object with `Alternative`s. However, they actually never have the fields `frames` & `polyline`. Not sure if you want to fix this.

bergmannjg commented:
@derhuerst I will change it in the next version 6.3.0.

traines-source commented:
@derhuerst Not sure I understand what your conclusion is. For the request params customization is already possible via `mapRouteParsers`, which will also adapt the OpenAPI schema, as used by [db-rest](https://github.com/derhuerst/db-rest/blob/6b6c1d6bdb6b41ea547d1fc6a59ac2d3a3e107e0/api.js#L89C1-L107C2). For the response schemas, this should already be possible via [`modifyRoutes`](https://github.com/derhuerst/db-rest/blob/6b6c1d6bdb6b41ea547d1fc6a59ac2d3a3e107e0/api.js#L109C1-L114C1), at least by inlining the modified/new schemas via the [`openapiPaths`](https://github.com/derhuerst/db-rest/blob/6b6c1d6bdb6b41ea547d1fc6a59ac2d3a3e107e0/routes/station.js#L26C14-L26C26) property.
